### ISSUE TITLE
Vim9: cannot refer to function variable in expression inside heredoc

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#10216]
**Steps to reproduce**

Run this shell command:

    vim -Nu NONE -S <(tee <<'EOF'
        vim9script
        def Func()
            var n = 123
            var msg =<< trim eval END
                n is equal to `=n`
            END
            echo msg[0]
        enddef
        Func()
    EOF
    )

An error is given at compile time:

    E121: Undefined variable: n

**Expected behavior**

No error is given at compile time.  This is echo'ed at runtime:

    n is equal to 123

**Version of Vim**

8.2 Included patches: 1-4779

**Environment**

Operating system: Ubuntu 20.04.4 LTS
Terminal: xterm
Value of $TERM: xterm-256color
Shell: GNU bash, version 5.0.17

**Additional context**

No issue if we move `n` from the function scope to the script one:
```vim
vim9script
var n = 123
def Func()
    var msg =<< trim eval END
        n is equal to `=n`
    END
    echo msg[0]
enddef
Func()
```
    n is equal to 123

---

No issue in legacy:
```vim
vim9script
function Func()
    let n = 123
    let msg =<< trim eval END
        n is equal to `=n`
    END
    echo msg[0]
endfunction
Func()
```
    n is equal to 123

---

[`:help :let-heredoc`](https://vimhelp.org/eval.txt.html#%3Alet-heredoc) says this:

   > Currenty, in a compiled function {expr} is evaluated
   > when compiling the function, THIS WILL CHANGE.

The `THIS WILL CHANGE` part seems to imply that evaluating the expression at compile time is undesirable, and that in the future it will be deferred to runtime.  But it doesn't say that we cannot refer to a function-local variable.  If this limitation is intended, maybe it should be documented.

**Comment by lacygoill** [comment]
> But it doesn't say that we cannot refer to a function-local variable.

Well, I guess it's implied.  At compile time, Vim probably cannot access the variables in the function-local scope yet.  But since it needs to evaluate the expression at compile time, it fails.

The issue will probably get fixed once the expression evaluation is deferred to runtime.  I still keep the issue open for the moment because I haven't found a TODO item for this.

**Comment by brammool** [comment]
I noticed this while including the changes.  Yegappan is looking into fixing this.
I think what you expect is that a local variable can be used to store the value you want, and then use that local variable (several times) in the heredoc text.  This is useful, thus we should make it work that way.

**Comment by lacygoill** [comment]
Probably fixed in patch [8.2.4804](https://github.com/vim/vim/releases/tag/v8.2.4804)
